### DISCUSSION TITLE
Complex block data structure support

### DISCUSSION THREAD

**Discussion by Reson-a** [toeverything/blocksuite#239] [Ideas]
Support complex nested data structure in collaborative block.
Only top level primitives are supported for now

Otherwise, we need to create lots of nested block in some cases. It's not necessary.

**Comment by doodlewind** [comment]
Hi, supporting the nested data structure is critical and should be the fundamental feature for `@blocksuite/store`. We envision that BlockSuite users can define complex models in a type-safe way, which could be used as something like [zod](https://github.com/colinhacks/zod). They could be implemented using a mechanism similar to [immer](https://github.com/immerjs/immer), automatically syncing standard model CRUD operations to underlying CRDT Yjs models. This could be a breakthrough in state management since the model built by this approach could support zero-cost time travel out of the box.

However, for now, our development resources need to focus on rich-text editing experience, so the comprehensive design of the store API may have to be delayed for (at least) about a month. I could post more about the draft ideas of the model API afterward, and you are welcome to share more thoughts if you are interested!
